packages/dds/merge-tree/src/perspective.ts

```diff
@@ -75,99 +92,158 @@ export class PerspectiveImpl implements Perspective {
 	 * @returns the previous segment, or the start of the tree if there is no previous segment.
 	 * @remarks This is a convenient equivalent to calling `nextSegment(segment, false)`.
 	 */
-	public previousSegment(segment: ISegmentLeaf): ISegmentLeaf {
-		return this.nextSegment(segment, false);
+	public previousSegment(mergeTree: MergeTree, segment: ISegmentLeaf): ISegmentLeaf {
+		return this.nextSegment(mergeTree, segment, false);
+	}
+
+	public isSegmentPresent(seg: ISegmentLeaf): boolean {
+		const insert: InsertOperationStamp = {
+			type: "insert",
+			clientId: seg.clientId,
+			seq: seg.seq,
+			localSeq: seg.localSeq,
+		};
+		if (isInserted(seg) && !this.hasOccurred(insert)) {
+			return false;
+		}
+
+		const removes: RemoveOperationStamp[] = [];
+		const removalInfo = toRemovalInfo(seg);
+		if (removalInfo !== undefined) {
+			removes.push(
+				...removalInfo.removedClientIds.map((clientId) =>
+					(clientId === LocalClientId || clientId === 0) &&
+					removalInfo.localRemovedSeq !== undefined
+						? ({
+								type: "setRemove",
+								seq: UnassignedSequenceNumber,
+								clientId,
+								localSeq: removalInfo.localRemovedSeq,
+							} as const)
+						: // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+							({ type: "setRemove", seq: removalInfo.removedSeq, clientId } as const),
+				),
+			);
+		}
+
+		const moveInfo = toMoveInfo(seg);
+		if (moveInfo !== undefined) {
+			removes.push(
+				...moveInfo.movedClientIds.map((clientId, index) =>
+					(clientId === LocalClientId || clientId === 0) &&
+					moveInfo.localMovedSeq !== undefined
+						? ({
+								type: "sliceRemove",
+								seq: UnassignedSequenceNumber,
+								clientId,
+								localSeq: moveInfo.localMovedSeq,
+							} as const)
+						: // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+							({ type: "setRemove", seq: moveInfo.movedSeqs[index]!, clientId } as const),
+				),
+			);
+		}
```

Comment:
note that the translation from the old in-memory representation to the new one we have to do here goes away and this method becomes a lot simpler :)